STYLE.md

```diff
@@ -0,0 +1,432 @@
+> sometimes lead to quadratic memory behavior when the nesting gets deep.
+
+### Function declaration
+
+All functions must have type signatures.
+
+Specialize function type signature for concrete types if you're using this function
+with only one type for each argument. Otherwise you should use more polymorphic
+version. Compiler can optimize specialized functions better
+and meaning of this function may be clearer.
+Use this rule unless you are the library creator and want your library to be abstract
+as possible.
+
+It is allowed to omit parentheses for only one type class constraint.
+
+If function type signature is very long then place type of each argument under
+its own line with respect to alignment.
+
+```haskell
+putValueInState
+  :: MonadIO m
+  => UserState
+  -> Maybe Int
+  -> AppConfig
+  -> (Int -> m ())
+  -> m ()
+```
+
+If the line with argument names is too big then put each argument on its own line
+and separate it somehow from body section.
+
+```haskell
+putValueInState
+  userState
+  mValue@(Just x)
+  appConfig
+  valueModificator
+  = do
+    <code goes here>
+```
+
+### List Declarations
+
+Align the elements in the list.  Example:
+
+```haskell
+exceptions =
+  [ InvalidStatusCode
+  , MissingContentHeader
+  , InternalServerError
+  ]
+```
+
+### If-then-else clauses
+
+Generally, guards and pattern matches should be preferred over _if-then-else_
+clauses, where possible.  Short cases should usually be put on a single line
+(when line length allows it).
+
+### Case expressions
+
+The alternatives in a case expression can be indented using either of
+the two following styles:
+
+```haskell
+foobar =
+  case something of
+    Just j  -> foo
+    Nothing -> bar
+```
+
+or as
+
+```haskell
+foobar = case something of
+           Just j  -> foo
+           Nothing -> bar
+```
+
+Align the `->` arrows when it helps readability.
+
+Dealing with laziness
+---------------------
+
+By default, use strict data types and lazy functions.
+
+### Data types
+
+Constructor fields should be strict, unless there's an explicit reason
+to make them lazy.  This helps to avoid many common pitfalls caused by too much
+laziness and reduces the number of brain cycles the programmer has to
+spend thinking about evaluation order.
+
+```haskell
+-- Good
+data Point = Point
+  { pointX :: !Double  -- ^ X coordinate
+  , pointY :: !Double  -- ^ Y coordinate
+  }
+```
+
+```haskell
+-- Bad
+data Point = Point
+  { pointX :: Double  -- ^ X coordinate
+  , pointY :: Double  -- ^ Y coordinate
+  }
+```
+
+TODO: Talk to luke
+Additionally, unpacking simple fields often improves performance and
+reduces memory usage:
+
+```haskell
+data Point = Point
+  { pointX :: {-# UNPACK #-} !Double  -- ^ X coordinate
+  , pointY :: {-# UNPACK #-} !Double  -- ^ Y coordinate
+  }
+```
+
+Misc
+----
+
+### Point-free style ###
+
+Avoid over-using point-free style.  For example, this is hard to read:
```

Comment:
:100:

Reply:
Now, I think we could make this more helpful, if we set some guidelines about what constitutes over-use.

Otherwise it's overwhelmingly left to judgement, and judgements can differ between reasonable people.

Reply:
> Now, I think we could make this more helpful, if we set some guidelines about what constitutes over-use.
> 
> Otherwise it's overwhelmingly left to judgement, and judgements can differ between reasonable people.

I think that's fair.

Reply:
Updated.

Reply:
> Now, I think we could make this more helpful, if we set some guidelines about what constitutes over-use.
> 
> Otherwise it's overwhelmingly left to judgement, and judgements can differ between reasonable people.

@deepfire: Do you have any opinions as to what would constitute over-use?

Reply:
@intricate, it's admittedly hard to put up useful (as in actionable) criteria. One thing that's actionable and definite, I think is use of composition sections -- `(foo .)`, at least I find that opaque enough.

Another one is less actionable, but rather -- if the expression is literally more complex (more elements) when point-free, then it's definitely not worth it.

Another one is when you have to use too many `flip`'s -- I'd say two is borderline too many.

But those are mostly soft criteria, admittedly.

You could say "the point-free version needs to have a clarity benefit" -- and that's easy to support, but then people have different views on clarity..

Reply:
I can't actually think of a place where I thought point free style was overused (or rather where it could not have been made perfectly fine with some better formatting).

Reply:
@erikd I would say that the same example in this document (a few lines below) is one that I would personally consider "bad" utilization of point-free:
```
f = (g .) . h
```
I find this unnecessarily hard to read.

Reply:
Would have to agree with @deepfire here.

Reply:
I will update with: "the point-free version needs to have a clarity benefit"  and resolve this. We can deal with on a case by case basis until we figure out actionable criteria.

Reply:
I agree `f = (g .) . h` is "overused pointfree".